mirgecom/steppers.py

```diff
@@ -93,3 +93,155 @@ def advance_state(rhs, timestepper, checkpoint, get_timestep,
             logmgr.tick_after()
 
     return istep, t, state
+
+
+def advance_state_leap(rhs, timestepper, checkpoint, get_timestep,
+                  state, t_final, component_id="state", t=0.0, istep=0,
+                  logmgr=None, eos=None, dim=None):
+    """Advance state from some time (t) to some time (t_final) using Leap.
+
+    Parameters
+    ----------
+    rhs
+        Function that should return the time derivative of the state
+    timestepper
+        Function that advances the state from t=time to t=(time+dt), and
+        returns the advanced state - in this case, a Leap method.
+    checkpoint
+        Function is user-defined and can be used to preform simulation status
+        reporting, viz, and restart i/o.  A non-zero return code from this function
+        indicates that this function should stop gracefully.
+    get_timestep
+        Function that should return dt for the next step. This interface allows
+        user-defined adaptive timestepping. A negative return value indicated that
+        the stepper should stop gracefully.
+    state: numpy.ndarray
+        Agglomerated object array containing at least the state variables that
+        will be advanced by this stepper
+    t_final: float
+        Simulated time at which to stop
+    component_id
+        State id (required input for leap method generation)
+    t: float
+        Time at which to start
+    istep: int
+        Step number from which to start
+
+    Returns
+    -------
+    istep: int
+        the current step number
+    t: float
+        the current time
+    state: numpy.ndarray
+    """
+    if t_final <= t:
+        return istep, t, state
+
+    # Generate code for Leap method.
+    dt = get_timestep(state=state)
+    code = timestepper.generate()
+    from dagrt.codegen import PythonCodeGenerator
+    codegen = PythonCodeGenerator(class_name="Method")
+    interp = codegen.get_class(code)(function_map={
+        "<func>" + component_id: rhs,
+        })
+    interp.set_up(t_start=t, dt_start=dt, context={component_id: state})
+
+    while t < t_final:
+
+        if logmgr:
+            logmgr.tick_before()
+
+        dt = get_timestep(state=state)
+        if dt < 0:
+            return istep, t, state
+
+        checkpoint(state=state, step=istep, t=t, dt=dt)
+
+        # Leap interface here is *a bit* different.
+        for event in interp.run(t_end=t+dt):
+            if isinstance(event, interp.StateComputed):
+                state = event.state_component
+                t += dt
+                istep += 1
+                if logmgr:
+                    set_dt(logmgr, dt)
+                    set_sim_state(logmgr, dim, state, eos)
+                    logmgr.tick_after()
+
+    return istep, t, state
+
+
+def advance(rhs, timestepper, checkpoint, get_timestep, state, t_final,
```

Comment:
Why not let this become `advance_state`? `_advance_state_callable` and `_advance_state_leap` themselves IMO don't need to be user-exposed.

Reply:
Done by 2caff69.